libs/isograph-react/src/core/startUpdate.ts

```diff
@@ -1,28 +1,323 @@
+  type ExtractUpdatableData,
   type FragmentReference,
   type UnknownTReadFromStore,
 } from './FragmentReference';
-import type { IsographEnvironment } from './IsographEnvironment';
+import {
+  assertLink,
+  type IsographEnvironment,
+  type Link,
+} from './IsographEnvironment';
+import { readPromise, type PromiseWrapper } from './PromiseWrapper';
+import {
+  readImperativelyLoadedField,
+  readLinkedFieldData,
+  readLoadablySelectedFieldData,
+  readResolverFieldData,
+  readScalarFieldData,
+  type NetworkRequestReaderOptions,
+  type ReadDataResultSuccess,
+} from './read';
+import type { ReaderAst } from './reader';
 
 export function getOrCreateCachedStartUpdate<
   TReadFromStore extends UnknownTReadFromStore,
 >(
   environment: IsographEnvironment,
-  fragmentReference: FragmentReference<TReadFromStore, any>,
+  fragmentReference: FragmentReference<TReadFromStore, unknown>,
   eagerResolverName: string,
+  networkRequestOptions: NetworkRequestReaderOptions,
 ): ExtractStartUpdate<TReadFromStore> {
   return (environment.eagerReaderCache[
     stableIdForFragmentReference(fragmentReference, eagerResolverName)
-  ] ??= createStartUpdate(environment, fragmentReference));
+  ] ??= createStartUpdate(
+    environment,
+    fragmentReference,
+    networkRequestOptions,
+  ));
 }
 
 export function createStartUpdate<TReadFromStore extends UnknownTReadFromStore>(
-  _environment: IsographEnvironment,
-  _fragmentReference: FragmentReference<TReadFromStore, any>,
+  environment: IsographEnvironment,
+  fragmentReference: FragmentReference<TReadFromStore, unknown>,
+  networkRequestOptions: NetworkRequestReaderOptions,
 ): ExtractStartUpdate<TReadFromStore> {
-  return (_updater) => {
-    // TODO start update
+  return (updater) => {
+    let mutableUpdatedIds: EncounteredIds = new Map();
+
+    let data = createUpdatableProxy(
+      environment,
+      fragmentReference,
+      networkRequestOptions,
+      mutableUpdatedIds,
+    );
+
+    updater(data);
+    callSubscriptions(environment, mutableUpdatedIds);
+  };
+}
+
+export function createUpdatableProxy<
+  TReadFromStore extends UnknownTReadFromStore,
+>(
+  environment: IsographEnvironment,
+  fragmentReference: FragmentReference<TReadFromStore, unknown>,
+  networkRequestOptions: NetworkRequestReaderOptions,
+  mutableUpdatedIds: EncounteredIds,
+): ExtractUpdatableData<TReadFromStore> {
+  const readerWithRefetchQueries = readPromise(
+    fragmentReference.readerWithRefetchQueries,
+  );
+
+  return readUpdatableData(
+    environment,
+    readerWithRefetchQueries.readerArtifact.readerAst,
+    fragmentReference.root,
+    fragmentReference.variables ?? {},
+    readerWithRefetchQueries.nestedRefetchQueries,
+    fragmentReference.networkRequest,
+    networkRequestOptions,
+    {
+      lastInvalidated: 0,
+    },
+    mutableUpdatedIds,
+  ).data;
+}
+
+function defineCachedProperty<T>(
+  target: T,
+  property: PropertyKey,
+  {
+    get,
+    set,
+    mutableState,
```

Comment:
Let's either make this function have a single parameter (an object with named fields) or five params. I don't like the half-way split here